src/maker.rs

```diff
@@ -0,0 +1,394 @@
+            self.dai_max_sell_amount.clone(),
+            self.mid_market_rate.value,
+            self.spread,
+        )
+    }
+
+    /// Decide whether we should proceed with order,
+    /// Confirm with the order book
+    /// Re & take & reserve
+    pub fn react_to_taken_order(&mut self, order: Order) -> anyhow::Result<Reaction> {
+        if self.ongoing_takers.cannot_trade_with_taker(&order.taker) {
+            return Ok(Reaction::CannotTradeWithTaker);
+        }
+
+        let current_profitable_rate = self
+            .spread
+            .apply(self.mid_market_rate.value, order.inner.position)?;
+        let order_rate = Rate::try_from(order.inner.clone())?;
+        let next_order = match order.clone().into() {
+            order
+            @
+            BtcDaiOrder {
+                position: Position::Buy,
+                ..
+            } => {
+                // We are buying BTC for DAI
+                // Given an order rate of: 1:9000
+                // It is NOT profitable to buy, if the current rate is greater than the order rate.
+                // 1:8800 -> We give less DAI for getting BTC -> Good.
+                // 1:9200 -> We have to give more DAI for getting BTC -> Sucks.
+                if order_rate > current_profitable_rate {
+                    return Ok(Reaction::RateSucks);
+                }
+
+                let updated_dai_reserved_funds = self.dai_reserved_funds.clone() + order.quote;
+                if updated_dai_reserved_funds > self.dai_balance {
+                    return Ok(Reaction::InsufficientFunds);
+                }
+
+                self.dai_reserved_funds = updated_dai_reserved_funds;
+                self.next_buy_order()?
+            }
+            order
+            @
+            BtcDaiOrder {
+                position: Position::Sell,
+                ..
+            } => {
+                // We are selling BTC for DAI
+                // Given an order rate of: 1:9000
+                // It is NOT profitable to sell, if the current rate is smaller than the order rate.
+                // 1:8800 -> We get less DAI for our BTC -> Sucks.
+                // 1:9200 -> We get more DAI for our BTC -> Good.
+                if order_rate < current_profitable_rate {
+                    return Ok(Reaction::RateSucks);
+                }
+
+                let updated_btc_reserved_funds = self.btc_reserved_funds + order.base;
+                if updated_btc_reserved_funds > self.btc_balance {
+                    return Ok(Reaction::InsufficientFunds);
+                }
+
+                self.btc_reserved_funds = updated_btc_reserved_funds;
+                self.next_sell_order()?
+            }
+        };
+
+        self.ongoing_takers
+            .insert(order.taker)
+            .expect("already checked that we can trade");
+
+        Ok(Reaction::Confirmed { next_order })
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Reaction {
```

Comment:
This enum is actually not needed, just use a `Result`.

Reply:
The enum helps to make things easier to understand and readable. Yes, it is technically not needed, but I think we should also distinguish an actual `Error` from a result that we expect (even though it is not the happy path).

I would tackle this in a next iteration.

Reply:
Are you sure we need to distinguish? :)

Reply:
It really helps when reading the tests as well... If you feel strong about it please go for follow up, I won't tackle it here. :)